restrict access to /admin to whitelisted ips

lunika opened the pull request:
## Purpose

We want to give access to /admin path to a restricted list of ips we
want to control. If the whitelist is empty the restriction is disabled and if ips are added to this list they will be used to restrict the access of /admin to this given list.

## Proposal

- [x] configure nginx on edxapp
- [x] configure nginx on richie
- [x] configure nginx on marsha

lunika committed:
✨(marsha) filter access to /admin with ips in a whitelist

We want to give access to /admin path to a restricted list of ips we
want to control. By default the variable marsha_nginx_admin_ip_withelist
is set with an empty list and if ips are added to this list they will be
used to restrict the access of /admin to this given list.

lunika committed:
✨(richie) filter access to /admin with ips in a whitelist

We want to give access to /admin path to a restricted list of ips we
want to control. By default the variable richie_nginx_admin_ip_withelist
is set with an empty list and if ips are added to this list they will be
used to restrict the access of /admin to this given list.

lunika committed:
✨(edxapp) filter access to /admin with ips in a whitelist

We want to give access to /admin path to a restricted list of ips we
want to control. By default the variable edxapp_nginx_admin_ip_withelist
is set with an empty list and if ips are added to this list they will be
used to restrict the access of /admin to this given list.

jmaupetit commented on apps/edxapp/templates/services/nginx/configs/lms.conf.j2:
```diff
@@ -44,6 +44,17 @@ server {
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+
+    if ($http_x_forwarded_for !~ ({{  edxapp_nginx_admin_ip_withelist | join("|") }})) {
```
```suggestion
    if ($http_x_forwarded_for !~ ({{ edxapp_nginx_admin_ip_withelist | join("|") }})) {
```

sampaccoud commented on CHANGELOG.md:
```diff
@@ -8,6 +8,10 @@ Versioning](http://semver.org/spec/v2.0.0.html).
 
+### Added
+
+- Whitelist ips authorized to acess /admin on edxapp, marsha and richie
```
```suggestion
- Whitelist ips authorized to access /admin on edxapp, marsha and richie
```

sampaccoud commented on CHANGELOG.md:
```diff
@@ -8,6 +8,10 @@ Versioning](http://semver.org/spec/v2.0.0.html).
+### Added
+
+- Whitelist ips authorized to acess /admin on edxapp, marsha and richie
+
```
Would be nice to apply it also to edxec's admin pages and to all pages of mailcatcher and learninglocker?

lunika replied:
done for learninglocker but for mailcatcher there is no nginx. I will check if the application allow to do that.

sampaccoud commented on apps/edxapp/templates/services/nginx/configs/cms.conf.j2:
```diff
@@ -34,6 +34,17 @@ server {
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+
+    if ($http_x_forwarded_for !~ ({{ edxapp_nginx_admin_ip_withelist | join("|") }})) {
```
I think `http_x_forwarded_for` is only set behind a proxy.

lunika replied:
You mean only if there is a haproxy/nginx/whatever in front of openshift ?

sampaccoud replied:
From what I understand, `HTTP_X_FORWARDED_FOR` can be a list of IPs when you are behind a number of proxies. If not behind a proxy, you should probably use `HTTP_X_REAL_IP`.

lunika replied:
But is it possible to directly expose an openshift or kubernetes cluster ? I think it's always behind a proxy

sampaccoud replied:
yes but even behind a proxy, `http_x_forwarded_for` is a list that grows and anybody can send whatever they want into it. We can use `http_x_real_ip` and specify that the openshift should be placed behind a proxy that is configured to set this header with the value of the real `remote_ip`?

lunika replied:
We have to use `http_x_forwarded_for` but keep in my mind that it's a comma separated list and the first occurence in the list is the real ip : https://github.com/symfony/http-foundation/blob/897bf05a4dd936cfd028d7d04f0f21590c024030/Request.php#L791-L798

lunika replied:
I made a modification. We know that the real ip is always the first one in the `http_x_forwarded_for` variable so we check if the string starts with the ip we want. WDYT ?

lunika committed:
✨(edxec) filter access to /admin with ips in a whitelist

We want to give access to /admin path to a restricted list of ips we
want to control. By default the variable edxec_nginx_admin_ip_withelist
is set with an empty list and if ips are added to this list they will be
used to restrict the access of /admin to this given list.

lunika committed:
✨(learninglocker) filter access to UI with ips in a whitelist

We want to give access to learninglocker UI to a restricted list of ips we
want to control. By default the variable learninglocker_nginx_admin_ip_withelist
is set with an empty list and if ips are added to this list they will be
used to restrict the access of the UI to this given list.

sampaccoud commented on apps/learninglocker/templates/services/nginx/configs/default.conf.j2:
```diff
@@ -42,6 +42,16 @@ server {
   location / {
+{% if learninglocker_nginx_admin_ip_withelist | length > 0 %}
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

sampaccoud commented on apps/richie/templates/services/nginx/configs/richie.conf.j2:
```diff
@@ -29,6 +29,20 @@ server {
+{% if richie_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

sampaccoud commented on apps/edxapp/templates/services/nginx/configs/cms.conf.j2:
```diff
@@ -34,6 +34,20 @@ server {
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

sampaccoud commented on apps/edxapp/templates/services/nginx/configs/lms.conf.j2:
```diff
@@ -44,6 +44,20 @@ server {
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

sampaccoud commented on apps/edxec/templates/services/nginx/configs/edxec.conf.j2:
```diff
@@ -29,6 +29,20 @@ server {
+{% if edxec_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

sampaccoud commented on apps/marsha/templates/services/nginx/configs/marsha.conf.j2:
```diff
@@ -29,6 +29,20 @@ server {
+{% if marsha_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

sampaccoud commented on CHANGELOG.md:
```diff
@@ -8,6 +8,12 @@ Versioning](http://semver.org/spec/v2.0.0.html).
+
+- Whitelist ips authorized to access /admin on edxapp,
+  edxec, marsha and richie
+- Whitelist ips authorized to access to learninglocker's UI
```
```suggestion
- Whitelist ips authorized to access learninglocker's UI
```

jmaupetit commented on apps/edxapp/templates/services/nginx/configs/cms.conf.j2:
```diff
@@ -34,6 +34,20 @@ server {
+  location /admin {
+    {# 
+      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in 
+      $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```
I suggest the following comment (I don't understand the current one):

>We want to limit access to a list of whitelisted IP addresses.
>
>`$http_x_forwarded_for` variable contains a list of IPs listed from the `HTTP_X_FORWARED_FOR` header (_e.g._ `w.x.y.z, 10.0.0.1`). The first IP corresponds to the client's public address, which is of interest (other ones have been added by subsequent proxies), hence we restrict our comparison with the beginning of this list (this is why our regex starts with a `^`).

jmaupetit commented on apps/edxapp/templates/services/nginx/configs/lms.conf.j2:
```diff
@@ -44,6 +44,20 @@ server {
     try_files $uri @proxy_to_lms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
```
Should we define two separated lists for the CMS and the LMS ?

lunika replied:
Yes I will do that

jmaupetit commented on apps/edxapp/templates/services/nginx/configs/cms.conf.j2:
```diff
@@ -34,6 +34,24 @@ server {
 
+{% if edxapp_nginx_cms_admin_ip_withelist | length > 0 %}
+  location /admin {
+        {# 
```
weird indentation

lunika replied:
good catch

jmaupetit reviewed (approved):
![](https://media0.giphy.com/media/jL6OeIhk3zPi/giphy.gif)
